src/mmap.rs

```diff
@@ -0,0 +1,583 @@
+// Copyright (C) 2019 Alibaba Cloud Computing. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+//! A default implementation of GuestMemory by mmap()-ing guest's memory into current process.
+//!
+//! The main structs to access guest's memory are:
+//! - MmapRegion: mmap a continuous region of guest's memory into current process
+//! - GuestRegionMmap: map from guest physical address into mmapped offset
+//! - GuestMemoryMmap: manage a collection of GuestRegionMmap objects
+
+use libc;
+use std::io::{self, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::ptr::null_mut;
+use std::sync::Arc;
+
+use address::Address;
+use guest_memory::*;
+use volatile_memory::{self, calc_offset, VolatileMemory, VolatileSlice};
+use Bytes;
+
+/// A backend driver to access guest's physical memory by mmapping guest's memory into current
+/// process.
+/// For a combination of 32-bit hypervisor and 64-bit virtual machine, only partial of guest's
+/// physical memory may be mapped into current process due to limited process virtual address
+/// space size.
+#[derive(Debug)]
+pub struct MmapRegion {
+    addr: *mut u8,
+    size: usize,
```

Comment:
How about adding another abstraction of AddressSpace to support the vhost crate?
The idea is to manage guest's physical address space with an AddressSpace object, which contains a group of address regions. Each address region maintains type(memory or MMIO), properties(RWX), Usage, GPA, optional fd, optional offset into fd etc.
And GuestMemory is a mechanism for current process to access the guest memory and it could be generated from the AddressSpace object on demand. 
Currently qemu maps all the guest memory into current process, but we have some thoughts only to map partial or even none of guest memory into the hypervisor process. For the vhost-user crate, we could simple generate a group of address regions accessible by IO subsystem (or specific device backend) and pass the (GPA, fd, offset) tuple to vhost-user backends.
What's your thoughts here?
Thanks!